src/borg/_chunker.c

```diff
@@ -236,6 +235,96 @@ chunker_fill(Chunker *c)
     return 1;
 }
 
+
+#ifdef __GNUC__
+#if __x86_64__
+/* All x86-64 CPUs support this, but how to invoke SSE2 mode might may vary by compiler, so we only activate
+ * that for GCC for now. */
+#define USE_SSE_ZERO
+#endif
+#endif
+
+#ifdef USE_SSE_ZERO
+
+static const int chunker_fast_zero = 1;
+
+#include <emmintrin.h>
+
+/* performance (Haswell, E3-1231v3): in cache about ~42 GB/s, w/ real memory access about 16-17 GB/s */
+
+__attribute__ ((target ("sse2")))
+static int
+chunker_is_zero(const uint8_t *data, ssize_t length)
+{
+    /* align */
+    unsigned align_diff = (0 - (uintptr_t) data) & 0xf;
+    if(align_diff) {
+        if(align_diff > length) {
+            /* length < 16 */
+            goto partial;
+        }
+        length -= align_diff;
+        for(; align_diff; data++, align_diff--) {
+            if(*data) {
+                return 0;
+            }
+        }
+    }
+
+    /* let's roll */
+    __m128i zerovec;
+    __m128i xmm0, xmm1, xmm2, xmm3;
+    zerovec = _mm_setzero_si128();
+
+    while((length -= 64) >= 0) {
+        xmm0 = _mm_load_si128((__m128i *)data);
+        xmm1 = _mm_load_si128((__m128i *)data + 1);
+        xmm2 = _mm_load_si128((__m128i *)data + 2);
+        xmm3 = _mm_load_si128((__m128i *)data + 3);
```

Comment:
i think writing the index explicitly is easier to parse (xmm0 = ... data, xmm1 = ... data + 1 etc.). Also allows to have the block size (64) explicit, which makes it clear that it has to be the same in all conditions.